[Impeller] Implement 'Picture::ToImage'

ColdPaleLight opened the pull request:
fix https://github.com/flutter/flutter/issues/109999

~We need to make `Picture::Snapshot` support the specified render target size, which will be used in the `ui.Picture.toImage` method in the future.~

Add 'Picture::ToImage' for `ui.Picture.toImage`

c.f.
https://api.flutter.dev/flutter/dart-ui/Picture/toImage.html

related issue
https://github.com/flutter/flutter/issues/104718


https://user-images.githubusercontent.com/31977171/186871536-aa7d3f2d-d25c-4a9a-a6f0-bc4bd743ad98.mp4



## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

bdero commented:
I'm a bit worried about using `Picture::Snapshot` for `Picture.toImage` (although that's partially because I disagree with the design of `Picture.toImage` in the first place). Up until now, coverage has been an internal implementation detail for optimizing Impeller, and using `Picture::Snapshot` for `Picture.toImage` would be our first leak this detail to users, thereby adding restrictions to what we can do with coverage in order to not break users.

Could you verify what happens when rendering a blurred rectangle using `Picture.toImage` with the Skia backend? Does the image get offset by the blur radius (so that the full blur appears in the top/left of the image? If this doesn't happen, a different utility needs to be written that doesn't consider coverage.

ColdPaleLight commented:
@bdero 
I get it, yes, we shouldn't expose coverage. I tweaked the code and added a new `Picture::ToImage` method for `ui.Picture.toImage`, which I verified in another draft PR, and it works fine.

I also tried `ImageFilter.blur`, and it doesn't change the offset on Skia backend (in the new implementation, I guess we can stop worrying about it.)

In the end I changed the Picture from Struct to Class as it started to get complicated.

Please let me know if you have any ideas. :)

bdero commented on impeller/aiks/picture.h:
```diff
@@ -15,10 +15,28 @@
+ private:
+  std::optional<impeller::Snapshot> DoSnapshot(AiksContext& context, Rect rect);
+
+  std::unique_ptr<EntityPass> pass_;
```
Since the setter and getter for this are trivial, can we just keep this public and remove them?

ColdPaleLight commented on impeller/aiks/picture.h:
```diff
@@ -15,10 +15,28 @@
+ public:
+  Picture() = default;
+
+  ~Picture() = default;
```
They have been removed.

ColdPaleLight commented on impeller/aiks/picture.cc:
```diff
@@ -49,7 +62,7 @@ std::optional<Snapshot> Picture::Snapshot(AiksContext& context) {
       .texture = std::move(texture),
-      .transform = translate.MakeTranslation(coverage->origin)};
-};
+      .transform = translate.MakeTranslation(rect.origin)};
```
Done.

ColdPaleLight commented on impeller/aiks/picture.cc:
```diff
@@ -15,22 +15,35 @@
+                              : nullptr;
+}
+
+std::optional<Snapshot> Picture::DoSnapshot(AiksContext& context, Rect rect) {
```
Done.

ColdPaleLight commented on impeller/aiks/picture.h:
```diff
@@ -15,10 +15,28 @@
+ private:
+  std::optional<impeller::Snapshot> DoSnapshot(AiksContext& context, Rect rect);
+
+  std::unique_ptr<EntityPass> pass_;
```
Done.

bdero reviewed (approved):
LGTM!
